File: frontend/src/api/eval.ts
/** 检索评测模块 REST 客户端（对齐后端 /v1/eval/* + schemas/eval.py，M27）。 */
import { api } from "./client";

// aggregate 的 recall/precision/ndcg 按 K，K 经 JSON 序列化为字符串 key（按 ["10"] 索引）。
export interface EvalAggregate {
  n: number;
  recall: Record<string, number | null>;
  precision: Record<string, number | null>;
  mrr: number | null;
  ndcg: Record<string, number | null>;
}

export interface EvalRunRequest {
  top_k?: number;
  rewrite?: "off" | "auto";
  eval_set?: string | null;
  persist?: boolean;
  ablation?: Record<string, boolean>; // M29: {"rerank": false} 跑单变体；省略=全开=生产
}

export interface EvalRunSummary {
  run_id: number;
  status: string;
  trigger: string;
  top_k: number;
  rewrite: string;
  embedding_strategy: string;
  n_queries: number;
  n_evaluable: number;
  rerank_on_count: number;
  duration_ms: number | null;
  unresolved_count: number;
  aggregate: EvalAggregate | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  error_message: string | null;
  ablation?: Record<string, boolean> | null; // M29: 单次评测变体配置（如 {"rerank": false}）；null=全开
}

export interface PerQueryRow {
  id: string;
  text: string;
  relevant?: string[];
  retrieved?: string[];
  missing?: string[];
  rerank_on: boolean;
  error: string | null;
  recall?: Record<string, number>;
  precision?: Record<string, number>;
  mrr?: number;
  ndcg?: Record<string, number>;
  first_hit_rank?: number | null;
  [k: string]: unknown;
}

export interface EvalRunDetail extends EvalRunSummary {
  config: Record<string, unknown> | null;
  per_query: PerQueryRow[] | null;
  unresolved: Array<Record<string, unknown>> | null;
}

export interface EvalRunListResponse {
  total: number;
  items: EvalRunSummary[];
}

// ===== A/B 消融（M28；对齐后端 schemas/eval.py AB* 模型）=====

export interface ABDelta {
  abs: number | null;
  pct: number | null;
}

export interface ABPairResult {
  name: string;
  claim: string;
  baseline: string;
  treatment: string;
  metric_focus: string[];
  // delta: { recall: {"10": {abs,pct}}, mrr: {abs,pct}, ... }
  delta: Record<string, Record<string, ABDelta> | ABDelta>;
}

export interface ABVariantResult {
  ablation: Record<string, boolean>;
  desc: string;
  aggregate: EvalAggregate | null;
  n_evaluable: number;
  n_queries: number;
  rerank_on_count: number;
  unresolved: number;
  per_query?: PerQueryRow[] | null;
}

export interface ABRunRequest {
  top_k?: number;
  rewrite?: "off" | "auto";
  eval_set?: string | null;
  pairs?: string[] | null; // ["rerank","multipath_rrf","graph"]；缺省=默认 3 组
  graph_subset?: boolean;
  diagnose?: boolean;
  persist?: boolean;
}

export interface ABRunSummary {
  run_id: number;
  status: string;
  trigger: string;
  top_k: number;
  rewrite: string;
  embedding_strategy: string;
  n_queries: number;
  n_evaluable: number;
  rerank_on_count: number;
  duration_ms: number | null;
  pairs: ABPairResult[];
  aggregate: EvalAggregate | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  error_message: string | null;
  kind: "ab";
}

export interface ABRunDetail extends ABRunSummary {
  variants: Record<string, ABVariantResult>;
  config: Record<string, unknown> | null;
}

export interface ABRunListResponse {
  total: number;
  items: ABRunSummary[];
}

// 评测 ~85 query × 全漏斗（含重排 API）≈ 数十秒，超过 client.ts 默认 30s 超时 → 单请求覆盖到 300s。
export const runEval = (body: EvalRunRequest) =>
  api.post<EvalRunDetail>("/v1/eval/run", body, { timeout: 300_000 }).then((r) => r.data);

export const listEvalRuns = (limit = 50, kind?: "single" | "ab") =>
  api
    .get<EvalRunListResponse>("/v1/eval/runs", { params: { limit, ...(kind ? { kind } : {}) } })
    .then((r) => r.data);

export const getEvalRun = (id: number) =>
  api.get<EvalRunDetail>(`/v1/eval/runs/${id}`).then((r) => r.data);

// A/B 跑 3~4 变体 × ~85 query × 全漏斗，比单次更慢 → 600s 超时。
export const runAb = (body: ABRunRequest) =>
  api.post<ABRunDetail>("/v1/eval/ab", body, { timeout: 600_000 }).then((r) => r.data);

export const listAbRuns = (limit = 50) =>
  api.get<ABRunListResponse>("/v1/eval/ab-runs", { params: { limit } }).then((r) => r.data);

export const getAbRun = (id: number, diagnose = false) =>
  api.get<ABRunDetail>(`/v1/eval/ab-runs/${id}`, { params: { diagnose } }).then((r) => r.data);

File: backend/app/schemas/eval.py
"""检索评测响应 schema（Phase 9 评测产品化 M27 / api接口清单 §eval）。

aggregate 的 recall/precision/ndcg 是按 K 的 map；K 经 JSONB 序列化为字符串 key
（``{"1":..,"10":..}``），故用 ``dict[str, float | None]``，前端按 ``["10"]`` 索引。
"""
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel


class EvalRunRequest(BaseModel):
    """``POST /v1/eval/run`` 请求。"""

    top_k: int = 10
    rewrite: Literal["off", "auto"] = "off"
    eval_set: str | None = None  # 缺省用 backend/eval/eval_set.yaml
    persist: bool = True  # False → 预览，不写库但仍返完整 detail
    ablation: dict[str, bool] | None = None  # M29: {"rerank": False, ...} 跑单变体；None=全开=生产；未知键→422


class EvalAggregate(BaseModel):
    """宏平均指标（``metrics.aggregate`` 形状）。"""

    n: int
    recall: dict[str, float | None]
    precision: dict[str, float | None]
    mrr: float | None
    ndcg: dict[str, float | None]


class EvalRunSummary(BaseModel):
    """历史列表项（不含 per_query，保持列表轻量）。"""

    run_id: int
    status: str
    trigger: str
    top_k: int
    rewrite: str
    embedding_strategy: str
    n_queries: int
    n_evaluable: int
    rerank_on_count: int
    duration_ms: int | None
    unresolved_count: int  # = len(unresolved)；>0 标红幽灵查询运行
    aggregate: EvalAggregate | None = None
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime
    error_message: str | None = None
    kind: str = "single"  # "single"(单次评测) | "ab"(A/B 消融，M28)；前端统一历史表的「类型」列
    ablation: dict[str, bool] | None = None  # M29: 单次评测跑变体时记录的消融配置（如 {"rerank": False}）；None=全开


class EvalRunDetail(EvalRunSummary):
    """单条详情（POST 运行结果 + GET /runs/{id}）：附 per_query / config / unresolved。"""

    config: dict[str, Any] | None = None
    per_query: list[dict[str, Any]] | None = None
    unresolved: list[dict[str, Any]] | None = None


class EvalRunListResponse(BaseModel):
    """``GET /v1/eval/runs`` 响应。"""

    total: int
    items: list[EvalRunSummary]


# ===== A/B 消融（M28；api接口清单 §eval-ab）=====
# ABReport（app/eval/ab_service.py）经 EvalRun.config["report"] 持久化（config.kind="ab" 区分）。


class ABRunRequest(BaseModel):
    """``POST /v1/eval/ab`` 请求。"""

    top_k: int = 10
    rewrite: Literal["off", "auto"] = "off"
    eval_set: str | None = None
    pairs: list[str] | None = None  # pair 名称子集（rerank/multipath_rrf/graph）；None=默认 3 组
    graph_subset: bool = False  # 对 call_chain 标签子集额外跑 graph pair
    diagnose: bool = False  # True → 详情返回向量路诊断字段（recall_paths/retrieved_kinds）
    persist: bool = True


class ABDelta(BaseModel):
    """单指标 baseline→treatment delta：abs=t−b；pct=(t−b)/b*100（b 为 None/0 → None）。"""

    abs: float | None = None
    pct: float | None = None


class ABPairResult(BaseModel):
    """一组 A/B 对照结果。``delta`` 形如 ``{recall:{1:{abs,pct},...}, mrr:{abs,pct}}``。"""

    name: str
    claim: str
    baseline: str
    treatment: str
    metric_focus: list[str]
    delta: dict[str, Any]


class ABVariantResult(BaseModel):
    """单个检索变体的聚合结果。per_query 仅在 detail（且 diagnose 时含诊断字段）出现。"""

    ablation: dict[str, Any]
    desc: str
    aggregate: EvalAggregate | None = None
    n_evaluable: int = 0
    n_queries: int = 0
    rerank_on_count: int = 0
    unresolved: int = 0
    per_query: list[dict[str, Any]] | None = None


class ABRunSummary(BaseModel):
    """A/B 历史列表项（不含 variants 的 per_query，保持列表轻量）。"""

    run_id: int
    status: str
    trigger: str
    top_k: int
    rewrite: str
    embedding_strategy: str
    n_queries: int
    n_evaluable: int
    rerank_on_count: int
    duration_ms: int | None
    pairs: list[ABPairResult]  # 各 pair 的 delta（轻量，列表即可看结论）
    aggregate: EvalAggregate | None = None  # full 变体锚点，供统一历史表趋势图
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime
    error_message: str | None = None
    kind: str = "ab"


class ABRunDetail(ABRunSummary):
    """A/B 详情（POST 运行结果 + GET /ab-runs/{id}）：附各变体明细 + 请求 config。"""

    variants: dict[str, ABVariantResult]
    config: dict[str, Any] | None = None


class ABRunListResponse(BaseModel):
    """``GET /v1/eval/ab-runs`` 响应。"""

    total: int
    items: list[ABRunSummary]


# ===== QA / 幻觉 eval（M39；config.kind="qa"，复用 eval_runs，零迁移）=====


class QARunRequest(BaseModel):
    """``POST /v1/eval/qa`` 请求。"""

    top_k: int = 8
    rewrite: Literal["off", "auto"] = "off"
    eval_set: str | None = None  # 缺省用 backend/eval/eval_set_qa.yaml
    persist: bool = True


class QADimensionScore(BaseModel):
    score: float | None = None
    weight: float = 1.0


class QAPerQueryRow(BaseModel):
    """单条 QA query 评判结果。"""

    id: str
    text: str
    answer: str
    citations_n: int = 0
    unverified_rate: float | None = None
    judge_scores: dict[str, Any] = {}  # {dim: float|None}
    rationale: str = ""
    weighted_score: float | None = None
    error: str | None = None


class QAAggregate(BaseModel):
    """QA 宏平均：5 维均值（4 judge + unverified_rate）+ weighted_quality（仅 rubric 4 维）。"""

    n: int
    means: dict[str, float | None]
    weighted_quality: float | None = None


class QARunSummary(BaseModel):
    """QA 历史列表项（不含 per_query）。"""

    run_id: int
    status: str
    trigger: str
    top_k: int
    rewrite: str
    embedding_strategy: str
    n_queries: int
    n_evaluable: int
    duration_ms: int | None
    aggregate: QAAggregate | None = None
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime
    error_message: str | None = None
    kind: str = "qa"


class QARunDetail(QARunSummary):
    """QA 详情：附 per_query + config。"""

    per_query: list[dict[str, Any]] | None = None
    config: dict[str, Any] | None = None


class QARunListResponse(BaseModel):
    total: int
    items: list[QARunSummary]
